### DISCUSSION TITLE
How to handle classroom.yml overwriting

### DISCUSSION THREAD

**Discussion by CodyHenrichsen-CTEC** [community/community#68249] [GitHub Education]
### Hi everyone,

I would love some assistance in using the work by @markpatterson27 for the pointsbar into the autograding, but the classroom.yml file I have in my template repo gets overwritten by github classroom. I thought I had seen a discussion on this but did not see it recently when searching.

How do avoid this? Is there an easy way? Or even a hard way?

Thank you!

**Comment by markpatterson27** [comment]
Hi Cody,

You need to include the test definitions `.github/classroom/autograding.json` file and `.github/workflows/classroom.yml` workflow file in the template repo that you are using for the assignment. These will be copied into the student's assignment repo when the student accepts the GH classroom assignment.

If you add any tests in the GH Classroom interface, GH Classroom will overwrite the custom workflow that is needed to update the [points bar](https://github.com/markpatterson27/points-bar).

![image](https://github.com/community/community/assets/35724907/0b69a502-4c2b-4692-8b7a-fd10e31143c1)

### Using Custom Workflow and Grading Tests

You can use the [example workflow](https://github.com/markpatterson27/points-bar#github-classroom-use) from the points bar README. This needs to be saved to `.github/workflows/classroom.yml` in the assignment starter code template.

You will need to manually write the `.github/classroom/autograding.json` file with your grading tests. This also goes into the assignment starter code template repo. The structure of the JSON is the same for each type of test (The GH Classroom interface has a choice of different tests, but they are all written using exactly the same JSON structure.) The `autograding.json` will look something like the following:

```json
{
    "tests": [
        {
            "name": "Test 1",
            "setup": "",
            "run": "command to run test 1'",
            "input": "",
            "output": "",
            "comparison": "included",
            "timeout": 10,
            "points": 5
        },
        {
            "name": "Test 2",
            "setup": "",
            "run": "command to run test 2",
            "input": "",
            "output": "",
            "comparison": "included",
            "timeout": 10,
            "points": 5
        }
    ]
}
```

I have a starter template for this file here: https://github.com/markpatterson27/GitHub-Classroom-Workflows/blob/main/.github/classroom/autograding.json

For each test:
`name`: name of the test. This can be anything.
`setup`: optional command to run before the test command
`run`: command to run the grading test
`input`: (optional) if the test command takes an input string.
`output`: (optional) if you are comparing the output from the grading test to this string value
`comparison`: string compare mode. Works with test output and expected `output` field.
`timeout`: int time in minutes before cancelling the test
`points`: (otional) int value for number of points given for passing the test

Leave any optional fields as an empty string if you aren't using that feature.

**Comment by CodyHenrichsen-CTEC** [comment]
Thank you soo much @markpatterson27 !!

**Comment by agilebotanist** [comment]
Very interesting discussion. I also look for ways to prevent students from altering the source files they are not supposed to touch.
I tried a very nice [Check-For-Changes action](https://github.com/markpatterson27/GitHub-Classroom-Check-For-Changes/) by @markpatterson27
It provides a custom check that fails on modification of specified files (btw, there is a bug there).

However, it won't help if students delete this action from the `classroom.yml` The check will stop working.

The only way I see is to overwrite those files for the final grading after the deadline and massively commit the fixed version for all students.

Any other ways?

Thanks in advance

**Reply by markpatterson27** [reply]
I forgot this even existed.
